Source/WTF/wtf/glib/SysprofAnnotator.h

```diff
@@ -0,0 +1,456 @@
+            return "RenderingUpdateRunLoopObserver";
+        case LayerTreeFreezeStart:
+        case LayerTreeFreezeEnd:
+            return "LayerTreeFreeze";
+        case FlushRemoteImageBufferStart:
+        case FlushRemoteImageBufferEnd:
+            return "FlushRemoteImageBuffer";
+        case CreateInjectedBundleStart:
+        case CreateInjectedBundleEnd:
+            return "CreateInjectedBundle";
+        case PaintSnapshotStart:
+        case PaintSnapshotEnd:
+            return "PaintSnapshot";
+        case RenderServerSnapshotStart:
+        case RenderServerSnapshotEnd:
+            return "RenderServerSnapshot";
+        case TakeSnapshotStart:
+        case TakeSnapshotEnd:
+            return "TakeSnapshot";
+        case SyntheticMomentumStart:
+        case SyntheticMomentumEnd:
+            return "SyntheticMomentum";
+        case SyntheticMomentumEvent:
+            return "SyntheticMomentumEvent";
+        case RemoteLayerTreeScheduleRenderingUpdate:
+            return "RemoteLayerTreeScheduleRenderingUpdate";
+
+        case CommitLayerTreeStart:
+        case CommitLayerTreeEnd:
+            return "CommitLayerTree";
+        case ProcessLaunchStart:
+        case ProcessLaunchEnd:
+            return "ProcessLaunch";
+        case InitializeSandboxStart:
+        case InitializeSandboxEnd:
+            return "InitializeSandbox";
+        case WebXRCPFrameWaitStart:
+        case WebXRCPFrameWaitEnd:
+            return "WebXRCPFrameWait";
+        case WebXRCPFrameStartSubmissionStart:
+        case WebXRCPFrameStartSubmissionEnd:
+            return "WebXRCPFrameStartSubmission";
+        case WebXRCPFrameEndSubmissionStart:
+        case WebXRCPFrameEndSubmissionEnd:
+            return "WebXRCPFrameEndSubmission";
+
+        case WakeUpAndApplyDisplayListStart:
+        case WakeUpAndApplyDisplayListEnd:
+            return "WakeUpAndApplyDisplayList";
+
+
+        case WTFRange:
+        case JavaScriptRange:
+        case WebCoreRange:
+        case WebKitRange:
+        case WebKit2Range:
+        case UIProcessRange:
+        case GPUProcessRange:
+        default:
+            return nullptr;
+        }
+    }
+
+    CString m_processName { "WebKit" };
+    static const char* getProcessName()
+    {
+        return getInstance().m_processName.data();
+    }
+
+    SysprofAnnotator() { };
+    ~SysprofAnnotator()
+    {
+        for (const auto& iter : ongoingMarks) {
+            std::pair<const void*, const void*> key = iter.first;
+            const void* pointer = std::get<0>(key);
+            const char* name = static_cast<const char*>(std::get<1>(key));
+            endMark(pointer, name, nullptr);
+        }
+    }
+
+    // This is actually <const void*, const char*>
+    std::map<std::pair<const void*, const void*>, std::pair<int64_t, std::string>> ongoingMarks;
+    Lock lock;
+
+    static SysprofAnnotator& getInstance()
+    {
+        static SysprofAnnotator instance;
```

Comment:
Hmm, who is destructing the instance? The destructor runs a non-trivial amount of code. The static object destruction order is not guaranteed - this looks suspicious to me.

Reply:
The destructor is never called, this is a forever object. I've removed this code from the destructor because I reasoned that Sysprof is about reporting marks and timestamps of concrete, finished events, and these events in the HashMap are not finished.

Reply:
It actually is going to be destroyed in an exit handler, which is risky. Please include wtf/NeverDestroyed.h and then use `NeverDestroyed<SysprofAnnotator>` to avoid this.